src/dialects/abstract/query-generator-typescript.ts

```diff
@@ -0,0 +1,111 @@
+import NodeUtil from 'node:util';
+import isObject from 'lodash/isObject';
+import type { ModelStatic } from '../../model.js';
+import type { Sequelize } from '../../sequelize.js';
+import { isPlainObject, isString } from '../../utils/index.js';
+import { isModelStatic } from '../../utils/model-utils.js';
+import type { TableName } from './query-interface.js';
+import type { AbstractDialect } from './index.js';
+
+export type TableNameOrModel = TableName | ModelStatic;
+
+export interface QueryGeneratorOptions {
+  sequelize: Sequelize;
+  dialect: AbstractDialect;
+}
+
+/**
+ * Temporary class to ease the TypeScript migration
+ */
+export class AbstractQueryGeneratorTypeScript {
+
+  protected readonly dialect: AbstractDialect;
+  protected readonly sequelize: Sequelize;
+
+  constructor(options: QueryGeneratorOptions) {
+    if (!options.sequelize) {
+      throw new Error('QueryGenerator initialized without options.sequelize');
+    }
+
+    if (!options.dialect) {
+      throw new Error('QueryGenerator initialized without options.dialect');
+    }
+
+    this.sequelize = options.sequelize;
+    this.dialect = options.dialect;
+  }
+
+  get options() {
+    return this.sequelize.options;
+  }
+
+  extractTableDetails(tableNameOrModel: TableNameOrModel, options?: { schema?: string, delimiter?: string }) {
+    const tableNameObject = isModelStatic(tableNameOrModel) ? tableNameOrModel.getTableName()
+      : isString(tableNameOrModel) ? { tableName: tableNameOrModel }
+      : tableNameOrModel;
+
+    if (!isPlainObject(tableNameObject)) {
+      throw new Error(`Invalid input received, got ${NodeUtil.inspect(tableNameOrModel)}, expected a Model Class, a TableNameWithSchema object, or a table name string`);
+    }
+
+    return {
+      ...tableNameObject,
+      schema: options?.schema || tableNameObject.schema || this.options.schema || this.dialect.getDefaultSchema(),
+      delimiter: options?.delimiter || tableNameObject.delimiter || '.',
+    };
+  }
+
+  /**
+   * Quote table name with optional alias and schema attribution
+   *
+   * @param param table string or object
+   * @param alias alias name
+   */
+  quoteTable(param: TableNameOrModel, alias: boolean | string = false): string {
+    if (isModelStatic(param)) {
+      param = param.getTableName();
+    }
+
+    const tableName = this.extractTableDetails(param);
+
+    if (isObject(param) && ('as' in param || 'name' in param)) {
+      throw new Error('parameters "as" and "name" are not allowed in the first parameter of quoteTable, pass them as the second parameter.');
+    }
+
+    if (alias === true) {
+      alias = tableName.tableName;
+    }
+
+    let sql = '';
+
+    if (this.dialect.supports.schemas) {
+      if (tableName.schema && tableName.schema !== this.dialect.getDefaultSchema()) {
```

Comment:
this is a change I'm unsure about.

Now that we always have a default schema per dialect, should we:

1. always add the schema to the query
2. or omit the schema if it's equal to the default schema, even if the user specified it manually?

Reply:
One of the reasons for omitting the schema if it's the default schema is that some users use the `schema` option in `sequelize.sync` to generate the same set of tables in different schemas, then use the `searchPath` option to execute code in different schemas: https://github.com/sequelize/sequelize/issues/5252#issuecomment-172666833

We'll make specifying the default schema equivalent to not specifying a schema, so this use case can continue